### PR TITLE
[persistence] Index the "ignored" anomaly property based on labels

### PR DESCRIPTION
**Pull request by shounakmk219**
#### Issue(s)

https://cortexdata.atlassian.net/browse/TE-1362

#### Description

Add an `ignored` field to the `MergedAnomalyResultIndex` using which anomalies can be filtered based on whether they are ignored. 
This field is computed by iterating over the anomaly labels and if there is a label with `ignore=true` then mark the anomaly as ignored.

### CODE REVIEW

File: thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/mapper/MergedAnomalyIndexMapper.java [hunk 1]
```diff
@@ -26,5 +29,19 @@ public interface MergedAnomalyIndexMapper {
 
   @Mapping(source = "enumerationItem.id", target = "enumerationItemId")
   @Mapping(target = "dimensions", ignore = true)
+  @Mapping(source = "anomalyLabels", target = "ignored", qualifiedByName = "labelsToIgnoredMapper")
   MergedAnomalyResultIndex toIndexEntity(AnomalyDTO dto);
+
+  @Named("labelsToIgnoredMapper")
+  static boolean labelsToIgnored(List<AnomalyLabelDTO> labels) {
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/mapper/MergedAnomalyIndexMapper.java:36]: tip: Use IntelliJ to simplify. this is a single liner
```
return labels != null && labels.stream().anyMatch(AnomalyLabelDTO::isIgnore);
```

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/AnomalyEntityTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import ai.startree.thirdeye.spi.datalayer.DaoFilter;
+import ai.startree.thirdeye.spi.datalayer.Predicate;
+import ai.startree.thirdeye.spi.datalayer.dto.AnomalyDTO;
+import ai.startree.thirdeye.spi.datalayer.dto.AnomalyLabelDTO;
+import java.util.List;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class AnomalyEntityTest {
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/AnomalyEntityTest.java:28]: This should be in the same `AnomalyManager` test.

**Review by suvodeep-pyne — CHANGES_REQUESTED** [review]

File: thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/mapper/MergedAnomalyIndexMapper.java (hunk 1 above)
Comment (shounakmk219) [reply thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/mapper/MergedAnomalyIndexMapper.java:36]: That's beautiful! Made the changes.

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/AnomalyEntityTest.java (hunk 2 above)
Comment (shounakmk219) [reply thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/AnomalyEntityTest.java]: Done

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/AnomalyEntityTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  private AnomalyDTO buildAnomaly() {
+    return new AnomalyDTO();
+  }
+
+  private AnomalyLabelDTO buildLabel(final boolean ignore) {
+    return new AnomalyLabelDTO().setIgnore(ignore);
+  }
+
+  @Test
+  public void ignoreLabelIndexTest() {
```
[the remark this answers is not in the document]
Comment (shounakmk219) [reply thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/AnomalyEntityTest.java]: Yes we do modification. Added new test cases.

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java [hunk 4]
```diff
@@ -531,4 +532,67 @@ private AnomalyDTO persist(final AnomalyDTO anomaly) {
     assertThat(id).isNotNull();
     return anomaly;
   }
+
+  @Test
+  public void ignoreLabelIndexNoLabelsTest() {
+    final Long id = mergedAnomalyResultDAO.save(buildAnomaly(0, 1));
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java:538]: just create another function here if you are using `buildAnomaly(0, 1)` and you don't care about the start and end time.

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java [hunk 5]
```diff
@@ -531,4 +532,67 @@ private AnomalyDTO persist(final AnomalyDTO anomaly) {
...
+    result.getAnomalyLabels().forEach(label -> label.setIgnore(false));
+    mergedAnomalyResultDAO.update(result);
+    assertByIdAndIgnored(id, false);
+  }
+
+  private AnomalyLabelDTO buildLabel(final boolean ignore) {
+    return new AnomalyLabelDTO().setIgnore(ignore);
+  }
+
+  private List<AnomalyDTO> filterByIdAndIgnored(final Long id, final boolean ignored) {
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java:586]: Is this a case where id is null ? Why `Long` and not `long`?

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java [hunk 6]
```diff
@@ -531,4 +532,67 @@ private AnomalyDTO persist(final AnomalyDTO anomaly) {
...
+    final AnomalyDTO result = mergedAnomalyResultDAO.findById(id);
+    result.getAnomalyLabels().get(0).setIgnore(true);
+    mergedAnomalyResultDAO.update(result);
+    assertByIdAndIgnored(id, true);
+  }
+
+  @Test
+  public void updateResetIgnoreLabelsTest() {
+    // create anomaly with ignore label
+    final List<AnomalyLabelDTO> labels = List.of(buildLabel(true), buildLabel(false));
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java:572]: nitpick: Let's just rename `buildLabel` to `label`. Ideally `buildAnomaly` should be just `anomaly` as well

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java [hunk 7]
```diff
@@ -531,4 +532,67 @@ private AnomalyDTO persist(final AnomalyDTO anomaly) {
...
+    final DaoFilter filter = new DaoFilter().setPredicate(Predicate.AND(
+        Predicate.EQ("baseId", id),
+        Predicate.EQ("ignored", ignored)
+    ));
+    return mergedAnomalyResultDAO.filter(filter);
+  }
+
+  private void assertByIdAndIgnored(final Long id, final boolean ignored) {
+    assertThat(filterByIdAndIgnored(id, ignored).size()).isEqualTo(1);
+    assertThat(filterByIdAndIgnored(id, !ignored).size()).isEqualTo(0);
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java:596]: Just for an additional check.

Let's also validate that the returned anomaly id should match the id as per the parameter.

**Review by suvodeep-pyne — DISMISSED** [review]
LGTM. minor comments. please fix and ship

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java (hunk 4 above)
Comment (shounakmk219) [reply thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java]: done

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java (hunk 5 above)
Comment (shounakmk219) [reply thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java:590]: changed it to `long`

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java (hunk 6 above)
Comment (shounakmk219) [reply thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java]: Renamed both the functions

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java (hunk 7 above)
Comment (shounakmk219) [reply thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/bao/TestAnomalyManager.java:602]: sure

**Pull request merged**